src/main/java/net/dries007/tfc/common/blockentities/BloomeryBlockEntity.java

```diff
@@ -0,0 +1,409 @@
+        for (int i = 0; i < 4; i++)
+        {
+            if (slagLayers > 0)
+            {
+                if (slagLayers >= 4)
+                {
+                    slagLayers -= 4;
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), TFCBlocks.MOLTEN.get().defaultBlockState().setValue(MoltenBlock.LIT, cooking).setValue(MoltenBlock.LAYERS, 4));
+                }
+                else
+                {
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), TFCBlocks.MOLTEN.get().defaultBlockState().setValue(MoltenBlock.LIT, cooking).setValue(MoltenBlock.LAYERS, slagLayers));
+                    slagLayers = 0;
+                }
+            }
+            else
+            {
+                //Remove any surplus slag(ie: after cooking/structure became compromised)
+                if (level.getBlockState(getInternalBlockPos().above(i)).is(TFCBlocks.MOLTEN.get()))
+                {
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), Blocks.AIR.defaultBlockState());
+                }
+            }
+        }
+    }
+
+    protected FluidStack getTotalInput()
+    {
+        Fluid inputFluid = null;
+        int totalInput = 0;
+        for (ItemStack stack : inputStacks)
+        {
+            HeatingRecipe heatingRecipe = HeatingRecipe.getRecipe(stack);
+            if (heatingRecipe != null)
+            {
+                if (inputFluid == null)
+                {
+                    inputFluid = heatingRecipe.getOutputFluid(new ItemStackInventory(stack)).getFluid();
+                }
+                if (inputFluid != null)
+                {
+                    FluidStack fluidStack = heatingRecipe.getOutputFluid(new ItemStackInventory(stack));
+                    if (inputFluid != fluidStack.getFluid())
+                    {
+                        throw new IllegalArgumentException("Bloomery had input items with different fluid types! That's not good!");
+                    }
+                    else
+                    {
+                        totalInput += fluidStack.getAmount();
+                    }
+                }
+            }
+        }
+
+        if (inputFluid == null)
+        {
+            return FluidStack.EMPTY;
+        }
+        else
+        {
+            return new FluidStack(inputFluid, totalInput);
+        }
+    }
+
+    protected ItemStack getTotalCatalyst()
+    {
+        if (catalystStacks.size() == 0)
+        {
+            return ItemStack.EMPTY;
+        }
+
+        int catalystCount = 0;
+        Item catalystItem = catalystStacks.get(0).getItem();
+        for (ItemStack stack : catalystStacks)
+        {
+            if (stack.getItem() == catalystItem)
+            {
+                catalystCount++;
+            }
+            else
+            {
+                throw new IllegalArgumentException("Bloomery had catalyst items of different types! That's not right!");
+            }
```

Comment:
```suggestion
```
This is crash code waiting to happen because someone drops an item on a bloomery wrong.